Google.Api.Generator/Utils/ResourcePattern.cs

```diff
@@ -16,107 +16,154 @@
-            public IReadOnlyList<string> ParameterNames { get; }
-            public IReadOnlyList<string> ParameterNamesWithSuffix { get; }
-            public int ParameterCount => ParameterNames.Count;
-            public bool IsComplex => ParameterCount > 1 || (ParameterCount == 1 && Separators.Any(x => x != ""));
+                void Check(bool cond, string msg)
+                {
+                    if (!cond)
+                    {
+                        throw new ArgumentException($"Segment '{segment}' is ill-formed; {msg}");
+                    }
+                }
+            }
 
-            public string Expand(IEnumerable<string> parameters) => Separators[0] + string.Join("", parameters.Zip(Separators.Skip(1), (p, s) => p + s));
+            // Separators are between parameter-names; there is always one less separator than parameter-names.
+            public override IReadOnlyList<char> Separators { get; }
+            public override IReadOnlyList<string> ParameterNames { get; }
+            public override IReadOnlyList<string> ParameterNamesWithSuffix { get; }
+            public override string PathTemplateString => $"{{{(ParameterCount == 1 ? ParameterNamesWithSuffix[0] : string.Join('_', ParameterNames))}}}";
+            public override string Expand(IEnumerable<string> parameters) =>
+                parameters.First() + string.Join("", Separators.Zip(parameters.Skip(1), (s, p) => $"{s}{p}"));
 
-            public string PathTemplateString => ParameterCount > 0 ? "{" + string.Join('_', ParameterNamesWithSuffix) + "}" : Separators.Single();
+            public override string ToString() =>
+                string.Join("", ParameterNamesWithSuffix.Zip(Separators, (p, s) => $"{{{p}}}{s}").Append($"{{{ParameterNamesWithSuffix[^1]}}}"));
         }
 
-        public ResourcePattern(string pattern) => Segments = pattern.Split('/').Select(x => new Segment(x)).ToList();
+        public ResourcePattern(string pattern) => Segments = pattern.Split('/').Select(Segment.Create).ToImmutableList();
+
+        private ResourcePattern(IEnumerable<Segment> segments) => Segments = segments.ToImmutableList();
 
         public IReadOnlyList<Segment> Segments { get; }
 
         public IEnumerable<string> ParameterNames => Segments.SelectMany(x => x.ParameterNames);
 
-        public string Expand(IEnumerable<string> parameters) => string.Join('/',
+        public string Expand(IReadOnlyList<string> parameters) => string.Join('/',
             Segments.Aggregate((result: ImmutableList<string>.Empty, paramOfs: 0),
                 (acc, seg) => (acc.result.Add(seg.Expand(parameters.Skip(acc.paramOfs))), acc.paramOfs + seg.ParameterCount),
                 acc => acc.result));
 
         public string PathTemplateString => string.Join('/', Segments.Select(x => x.PathTemplateString));
+
+        public string ParentComparisonString => string.Join('/', Segments.Select(seg => seg switch
```

Comment:
Comment for what this is about, with examples?
Basically this allows `projects/{project}` and `projects/{something_else}` to match each other?

Reply:
Done